src/zenml/zen_server/utils.py

```diff
@@ -318,6 +337,150 @@ def decorated(*args: Any, **kwargs: Any) -> Any:
     return cast(F, decorated)
 
 
+class RequestLimiter:
+    """Simple in-memory rate limiter."""
+
+    def __init__(
+        self,
+        day_limit: Optional[int] = None,
+        minute_limit: Optional[int] = None,
+    ):
+        """Initializes the limiter.
+
+        Args:
+            day_limit: The number of requests allowed per day.
+            minute_limit: The number of requests allowed per minute.
+
+        Raises:
+            ValueError: If both day_limit and minute_limit are None.
+        """
+        self.limiting_enabled = server_config().rate_limit_enabled
+        if not self.limiting_enabled:
+            return
+        if day_limit is None and minute_limit is None:
+            raise ValueError("Pass either day or minuter limits, or both.")
+        self.day_limit = day_limit
+        self.minute_limit = minute_limit
+        self.limiter: Dict[str, List[float]] = defaultdict(list)
+
+    def hit_limiter(self, request: Request) -> None:
+        """Increase the number of hits in the limiter.
+
+        Args:
+            request: Request object.
+
+        Raises:
+            HTTPException: If the request limit is exceeded.
+        """
+        if not self.limiting_enabled:
+            return
+        from fastapi import HTTPException
+
+        requester = self._get_ipaddr(request)
+        now = time.time()
+        minute_ago = now - 60
+        day_ago = now - 60 * 60 * 24
+        self.limiter[requester].append(now)
+
+        # remove failures older than a day
+        older_index = None
+        for i, limiter_hit in enumerate(self.limiter[requester]):
+            if limiter_hit < day_ago:
+                older_index = i
+            else:
+                break
+        if older_index is not None:
+            self.limiter[requester] = self.limiter[requester][
+                older_index + 1 :
+            ]
```

Comment:
```suggestion
        from bisect import bisect_left
        # remove failures older than a day
        older_index = bisect_left(self.limiter[requester], day_ago)
        self.limiter[requester] = self.limiter[requester][older_index:]
```

With large sorted lists, you can leverage a binary search algorithm to reduce the time to find the delimitation.